File: packages/api/src/ui/server/system.ts
import { checkPermission } from "tauri-api-adapter/permissions"
import {
	ejectAllDisks,
	emptyTrash,
	getFrontmostApp,
	getSelectedFilesInFileExplorer,
	hideAllAppsExceptFrontmost,
	logoutUser,
	mute,
	openTrash,
	quitAllApps,
	reboot,
	setVolume,
	setVolumeTo0,
	setVolumeTo25,
	setVolumeTo50,
	setVolumeTo75,
	setVolumeTo100,
	showDesktop,
	shutdown,
	sleep,
	sleepDisplays,
	toggleBluetooth,
	toggleHiddenFiles,
	toggleMute,
	toggleStageManager,
	toggleSystemAppearance,
	turnVolumeDown,
	turnVolumeUp,
	unmute
} from "../../commands/system"
import {
	AllKunkunPermission,
	type FsPermissionScoped,
	type KunkunFsPermission,
	type SystemPermission
} from "../../permissions"
import { SystemPermissionMap } from "../../permissions/permission-map"
import type { ISystem } from "../client"

export function constructSystemApi(permissions: SystemPermission[]): ISystem {
	return {
		openTrash: checkPermission<SystemPermission>(
			SystemPermissionMap.openTrash,
			permissions
		)(openTrash),
		emptyTrash: checkPermission<SystemPermission>(
			SystemPermissionMap.emptyTrash,
			permissions
		)(emptyTrash),
		shutdown: checkPermission<SystemPermission>(
			SystemPermissionMap.shutdown,
			permissions
		)(shutdown),
		reboot: checkPermission<SystemPermission>(SystemPermissionMap.reboot, permissions)(reboot),
		sleep: checkPermission<SystemPermission>(SystemPermissionMap.sleep, permissions)(sleep),
		toggleSystemAppearance: checkPermission<SystemPermission>(
			SystemPermissionMap.toggleSystemAppearance,
			permissions
		)(toggleSystemAppearance),
		showDesktop: checkPermission<SystemPermission>(
			SystemPermissionMap.showDesktop,
			permissions
		)(showDesktop),
		quitAllApps: checkPermission<SystemPermission>(
			SystemPermissionMap.quitAllApps,
			permissions
		)(quitAllApps),
		sleepDisplays: checkPermission<SystemPermission>(
			SystemPermissionMap.sleepDisplays,
			permissions
		)(sleepDisplays),
		setVolume: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolume,
			permissions
		)(setVolume),
		setVolumeTo0: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolumeTo0,
			permissions
		)(setVolumeTo0),
		setVolumeTo25: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolumeTo25,
			permissions
		)(setVolumeTo25),
		setVolumeTo50: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolumeTo50,
			permissions
		)(setVolumeTo50),
		setVolumeTo75: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolumeTo75,
			permissions
		)(setVolumeTo75),
		setVolumeTo100: checkPermission<SystemPermission>(
			SystemPermissionMap.setVolumeTo100,
			permissions
		)(setVolumeTo100),
		turnVolumeUp: checkPermission<SystemPermission>(
			SystemPermissionMap.turnVolumeUp,
			permissions
		)(turnVolumeUp),
		turnVolumeDown: checkPermission<SystemPermission>(
			SystemPermissionMap.turnVolumeDown,
			permissions
		)(turnVolumeDown),
		toggleStageManager: checkPermission<SystemPermission>(
			SystemPermissionMap.toggleStageManager,
			permissions
		)(toggleStageManager),
		toggleBluetooth: checkPermission<SystemPermission>([], permissions)(toggleBluetooth),
		toggleHiddenFiles: checkPermission<SystemPermission>(
			SystemPermissionMap.toggleHiddenFiles,
			permissions
		)(toggleHiddenFiles),
		ejectAllDisks: checkPermission<SystemPermission>(
			SystemPermissionMap.ejectAllDisks,
			permissions
		)(ejectAllDisks),
		logoutUser: checkPermission<SystemPermission>(
			SystemPermissionMap.logoutUser,
			permissions
		)(logoutUser),
		toggleMute: checkPermission<SystemPermission>(
			SystemPermissionMap.toggleMute,
			permissions
		)(toggleMute),
		mute: checkPermission<SystemPermission>(SystemPermissionMap.mute, permissions)(mute),
		unmute: checkPermission<SystemPermission>(SystemPermissionMap.unmute, permissions)(unmute),
		getFrontmostApp: checkPermission<SystemPermission>(
			SystemPermissionMap.getFrontmostApp,
			permissions
		)(getFrontmostApp),
		hideAllAppsExceptFrontmost: checkPermission<SystemPermission>(
			SystemPermissionMap.hideAllAppsExceptFrontmost,
			permissions
		)(hideAllAppsExceptFrontmost),
		getSelectedFilesInFileExplorer: checkPermission<SystemPermission>(
			SystemPermissionMap.getSelectedFilesInFileExplorer,
			permissions
		)(getSelectedFilesInFileExplorer)
	}
}

// all enabled by default
// export const defaultSystemApi = constructSystemApi([
// 	"system:volumn",
// 	"system:boot",
// 	"system:disk",
// 	"system:apps",
// 	"system:fs"
// ])

File: packages/tauri-plugins/jarvis/src/commands/system.rs
use crate::syscmds::{CommonSystemCmds, SystemCmds};
use crate::utils::script::run_apple_script;
use applications::AppInfo;

#[tauri::command]
pub async fn open_trash() -> Result<(), String> {
    SystemCmds::open_trash().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn empty_trash() -> Result<(), String> {
    SystemCmds::empty_trash().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn shutdown() -> Result<(), String> {
    SystemCmds::shutdown().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn reboot() -> Result<(), String> {
    SystemCmds::reboot().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn sleep() -> Result<(), String> {
    SystemCmds::sleep().map_err(|err| err.to_string())
}

/// ```applescript
/// tell application "System Events" to tell appearance preferences
///   set dark_mode to dark mode
///   set dark_mode to not dark_mode
///   set dark mode to dark_mode
/// end tell
/// ```
#[tauri::command]
pub async fn toggle_system_appearance() -> Result<(), String> {
    run_apple_script(
        r#"
        tell application "System Events" to tell appearance preferences
            set dark_mode to dark mode
            set dark_mode to not dark_mode
            set dark mode to dark_mode
        end tell
    "#,
    )
    .map_err(|err| err.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn show_desktop() -> Result<(), String> {
    run_apple_script("tell application \"System Events\" to key code 103")
        .map_err(|err| err.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn quit_all_apps() -> Result<(), String> {
    todo!()
}

#[tauri::command]
pub async fn sleep_displays() -> Result<(), String> {
    run_apple_script("do shell script \"pmset displaysleepnow\"").map_err(|err| err.to_string())?;
    Ok(())
}

/// Set Volume to 0
#[tauri::command]
pub async fn set_volume(percentage: u8) -> Result<(), String> {
    SystemCmds::set_volume(percentage).map_err(|err| err.to_string())
}
/// Turn Volume Up
#[tauri::command]
pub async fn turn_volume_up() -> Result<(), String> {
    SystemCmds::turn_volume_up().map_err(|err| err.to_string())
}
/// Turn Volume Down
#[tauri::command]
pub async fn turn_volume_down() -> Result<(), String> {
    SystemCmds::turn_volume_down().map_err(|err| err.to_string())
}
/// Toggle Stage Manager
#[tauri::command]
pub async fn toggle_stage_manager() -> Result<(), String> {
    todo!()
}

/// Toggle Bluetooth
#[tauri::command]
pub async fn toggle_bluetooth() -> Result<(), String> {
    todo!()
}

/// Toggle Hidden Files
#[tauri::command]
pub async fn toggle_hidden_files() -> Result<(), String> {
    todo!()
}

/// Eject All Disks
#[tauri::command]
pub async fn eject_all_disks() -> Result<(), String> {
    run_apple_script("tell application \"Finder\" to eject (every disk whose ejectable is true)")
        .map_err(|err| err.to_string())?;
    Ok(())
}
/// Logout <User>
#[tauri::command]
pub async fn logout_user() -> Result<(), String> {
    SystemCmds::logout_user().map_err(|err| err.to_string())
}
/// Toggle Mute
#[tauri::command]
pub async fn toggle_mute() -> Result<(), String> {
    SystemCmds::toggle_mute().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn mute() -> Result<(), String> {
    SystemCmds::mute().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn unmute() -> Result<(), String> {
    SystemCmds::unmute().map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn get_frontmost_app() -> Result<applications::App, String> {
    let ctx = applications::AppInfoContext::new();
    ctx.get_frontmost_application()
        .map_err(|err| err.to_string())
}

/// Hide All Apps Except Frontmost
#[tauri::command]
pub async fn hide_all_apps_except_frontmost() -> Result<(), String> {
    run_apple_script(
        r#"tell application "System Events"
            set frontApp to name of first application process whose frontmost is true
            set visibleApps to every process whose visible is true and name is not frontApp
            repeat with theApp in visibleApps
                set visible of theApp to false
            end repeat
        end tell"#,
    )
    .map_err(|err| err.to_string())?;
    Ok(())
}

/// For macOS, return the selected files in the finder
#[tauri::command]
pub async fn get_selected_files_in_file_explorer() -> Result<Vec<std::path::PathBuf>, String> {
    SystemCmds::get_selected_files().map_err(|err| err.to_string())
}
